src/linear_algebra/quadratic_form.lean

```diff
@@ -259,6 +261,25 @@ by simp [sub_eq_add_neg]
 @[simp] lemma sub_apply (Q Q' : quadratic_form R M) (x : M) : (Q - Q') x = Q x - Q' x :=
 by simp [sub_eq_add_neg]
 
+/-- `@coe_fn (quadratic_form R M)` as an `add_monoid_hom`. -/
+@[simps apply]
+def coe_fn_add_monoid_hom : quadratic_form R M →+ (M → R) :=
+{ to_fun := coe_fn, map_zero' := coe_fn_zero, map_add' := coe_fn_add }
```

Comment:
I don't think this is the right definition:
 * Since `quadratic_form R M` is also a module over `R`, shouldn't this be a linear map?
 * I would call this definition `eval` (or `apply`), not `coe_fn`, since "evaluating a quadratic form" is a mathematical concept, while `coe_fn` is a Lean-specific (mathlib-specific) concept. This also matches with `monoid_hom.eval`, `dual.eval`, arguably `polynomial.eval`.
 * Should the order of the `quadratic_form` and `M` parameters be flipped, to match with `monoid_hom.eval`? (I don't think so, since the map is not linear/additive in the `M` parameter.)

Reply:
This isn't the evaluation map though, this is the coercion map analogous to `monoid_hom.eval`.

Reply:
The evaluation map anaglous to `monoid_hom.eval` is _not_ additive, as obviously it's quadratic!

Reply:
> Since quadratic_form R M is also a module over R, shouldn't this be a linear map?

It could be, but it would be linear over some `S` where `[comm_semiring S] [algebra S R]`, as `R` may not be commutative.
While possibly useful to someone, that's hard to work with when all you want is a `→+`.

Reply:
> This isn't the evaluation map though, this is the coercion map analogous to [`monoid_hom.coe_fn`](https://leanprover-community.github.io/mathlib_docs/algebra/group/pi.html#monoid_hom.coe_fn) and [`coe_fn_add_monoid_hom`](https://leanprover-community.github.io/mathlib_docs/topology/continuous_function/algebra.html#coe_fn_add_monoid_hom) (which should really be in a namespace!)

I see, we have established a convention that `eval = flip coe_fn`. Not my favourite idea, but I can live with it.

Could you at least add the word "evaluation" in the docstring?

Reply:
I can make the docstring a little clearer as:
```suggestion
/-- `@coe_fn (quadratic_form R M)` as an `add_monoid_hom`.

That is, coercion of a `quadratic_form` into a function is an add_monoid_hom. -/
@[simps apply]
def coe_fn_add_monoid_hom : quadratic_form R M →+ (M → R) :=
{ to_fun := coe_fn, map_zero' := coe_fn_zero, map_add' := coe_fn_add }
```
There's no evaluation happening anywhere here, so I don't see how to get that word into the docstring.

Reply:
> * while `coe_fn` is a Lean-specific (mathlib-specific) concept.

This is kind of the point in the definition. Lean / mathlib has lots of concepts of function, and this statement is just "switching between two of those concepts doesn't affect addition".

Reply:
> > Since quadratic_form R M is also a module over R, shouldn't this be a linear map?
> 
> It could be, but it would be linear over some `S` where `[comm_semiring S] [algebra S R]`, as `R` may not be commutative.
> While possibly useful to someone, that's hard to work with when all you want is a `→+`.

It would fit with the [bundled maps over different rings pattern](https://leanprover-community.github.io/mathlib_docs/notes.html#bundled%20maps%20over%20different%20rings): define it for all `S` and instantiate `S := ℕ` when concluding the sum of q.f.s is the q.f. given by the sum.

Reply:
> This is kind of the point in the definition. Lean / mathlib has lots of concepts of function, and this statement is just "switching between two of those concepts is an additive map". It's not expected to be used mathematically anyway, it's just glue for mathlib APIs.

I disagree that a quadratic form is just another kind of function, at least in my fields of interest. I have seen at least as many instances where a homogeneous polynomial or a matrix is taken as the "underlying" representation, and I wouldn't argue that `polynomial.aeval` is just glue for mathlib APIs. (The current mathlib definition is still nicer to formalize with, though!)

Reply:
Also, the docstring begins with "A quadratic form is a map `Q : M → R` "!

Reply:
It looks like I forgot to send my earlier comment regarding linearity:

> It would fit with the bundled maps over different rings pattern

True; but my claim is that no one ever needs the linear version of `coe_fn`, based on the fact I haven't seen it defined for any other types. The additive and multiplicative versions are useful because they bundle proofs about `pow`, `finset.sum/prod`, `finsupp.sum`, and `dfinsupp.sum`. I haven't run into any lemmas on `linear_map` that are useful in this way. Based on that, I'd prefer to declare adding it out of scope.

> define it for all S and instantiate S := ℕ when concluding the sum of q.f.s is the q.f. given by the sum.

If you do that you also end up having to wait until you have a semimodule structure available, meaning this would have to move down the file, and the `coe_fn_sum` lemmas would no longer be next to the other pre-smul stuff.